### ISSUE TITLE
Feature: Date column is incorrectly detected when copying from Pandas DataFrame

### ISSUE THREAD

**Issue by prrao87** [kuzudb/kuzu#3855]
### API

Python

### Description

Tested on 0.4.3.dev57.

There's some weird behaviour when copying from a Pandas DataFrame with a date column.
```
id,transaction_date,is_fraudulent
1,2023-01-31,true
2,2021-03-17,false
3,2024-02-29,true
```

## Kùzu

```py
import shutil
import kuzu
import pandas as pd

shutil.rmtree("./testdb", ignore_errors=True)
db = kuzu.Database("./testdb")
conn = kuzu.Connection(db)

conn.execute(
    """
    CREATE NODE TABLE Transaction(
        id INT64,
        transaction_date DATE,
        is_fraudulent BOOLEAN,
        PRIMARY KEY(id)
    )
    """
)

df = pd.read_csv("test.csv", parse_dates=["transaction_date"])
print(df)

# Read into Kuzu
conn.execute("COPY Transaction FROM df")
```
Error:
```
   id transaction_date  is_fraudulent
0   1       2023-01-31           True
1   2       2021-03-17          False
2   3       2024-02-29           True
Traceback (most recent call last):
  File "/Users/prrao/code/kuzu-debug/dataframes/test.py", line 24, in <module>
    conn.execute("COPY Transaction FROM df")
  File "/Users/prrao/code/kuzu-debug/.venv/lib/python3.12/site-packages/kuzu/connection.py", line 130, in execute
    _query_result = self._connection.query(query)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Binder exception: transaction_date has data type TIMESTAMP_NS but DATE was expected.
```

## DuckDB

It just works, no effort required.

```python
import duckdb
import pandas as pd

conn = duckdb.connect()

df = pd.read_csv("test.csv", parse_dates=["transaction_date"])
print(df)

conn.execute(
    """
    CREATE TABLE Transaction(
        id INT64,
        transaction_date DATE,
        is_fraudulent BOOLEAN
    )
    """
)
# Read into DuckDB
conn.execute("INSERT INTO Transaction SELECT * FROM df")

res = conn.execute("SELECT * FROM Transaction")
print(res.pl())
```
```
   id transaction_date  is_fraudulent
0   1       2023-01-31           True
1   2       2021-03-17          False
2   3       2024-02-29           True
shape: (3, 3)
┌─────┬──────────────────┬───────────────┐
│ id  ┆ transaction_date ┆ is_fraudulent │
│ --- ┆ ---              ┆ ---           │
│ i64 ┆ date             ┆ bool          │
╞═════╪══════════════════╪═══════════════╡
│ 1   ┆ 2023-01-31       ┆ true          │
│ 2   ┆ 2021-03-17       ┆ false         │
│ 3   ┆ 2024-02-29       ┆ true          │
└─────┴──────────────────┴───────────────┘
```

## Desired behaviour

The date column should be treated differently from a timestamp column when copying from Pandas.

**Comment by mxwli** [comment]
The way COPY FROM is implemented for pandas, pyarrow, and polars is by directly copying scan output to destination tables. This requires that the types be matched exactly, which seems to be the reason your example fails. The best solution right now would probably be to apply casting after scanning; @andyfengHKU should know more about this.

**Comment by prrao87** [comment]
But why does this work in Polars, but not pandas? I think it's a numpy data type problem, because the Arrow date object under the hood doesn't cause an issue in Polars. Does numpy not have a `date` object that's not a timestamp?

**Comment by prrao87** [comment]
Also, another reason why I think some level of implicit casting is helpful is that DuckDB is able to handle the same data without requiring manual casting. And dates are an incredibly common data type that folks works with in Pandas.

**Comment by andyfengHKU** [comment]
I have implemented the implicit casting in #3887 and #3891. Though I'm still facing the same problem. 

From the internal perspective, I get an input type `TIMESTAMP` and try to cast it to `DATE`. This cannot be done as an implicit casting yet because we might lose the time part. 

There are two solutions

1. implement implicit cast from `TIMESTAMP` to `DATE` and throw if we are going to lose time,
2. directly interpret the column as `DATE`

@mxwli should first check if 2 make sense. If not, we should go with 1.

**Comment by prrao87** [comment]
I'm think 1 makes sense? If the user specified the DATE type for the property when creating the Kuzu node table, they have "opted in" to throwing away the time. If they wanted the time then they should create the column of that type during table creation.

**Comment by mxwli** [comment]
Implicit casting from TIMESTAMP to DATE is valid. Option 2 doesn't really make sense to me. It seems that numpy reads the date column into datetime. This works in Polars because arrow supports pure date types.

**Comment by prrao87** [comment]
Yup, let's go with option 1 then. We should support it as Pandas can handle pure dates, and it's seamlessly read into DuckDB too.

**Comment by mxwli** [comment]
Works now. Closed by #4041
